pkg/vmprovider/providers/vsphere2/virtualmachine/backup.go

```diff
@@ -28,136 +42,174 @@ type VMDiskData struct {
-	if bootstrapDataBackup == "" {
-		ctx.VMCtx.Logger.V(4).Info("Skipping VM bootstrap data backup as unchanged")
+	if instanceID == "" {
+		opts.VMCtx.Logger.V(4).Info("Skipping cloud-init instance ID as already stored")
 	} else {
 		ecToUpdate = append(ecToUpdate, &types.OptionValue{
-			Key:   vmopv1.VMBackupBootstrapDataExtraConfigKey,
-			Value: bootstrapDataBackup,
+			Key:   vmopv1.CloudInitInstanceIDExtraConfigKey,
+			Value: instanceID,
 		})
 	}
 
-	diskDataBackup, err := getDesiredDiskDataForBackup(ctx, curEcMap)
+	pvcDiskData, err := getDesiredPVCDiskDataForBackup(opts, curEcMap)
 	if err != nil {
-		ctx.VMCtx.Logger.Error(err, "Failed to get VM disk data for backup")
+		opts.VMCtx.Logger.Error(err, "failed to get PVC disk data for backup")
 		return err
 	}
 
-	if diskDataBackup == "" {
-		ctx.VMCtx.Logger.V(4).Info("Skipping VM disk data backup as unchanged")
+	if pvcDiskData == "" {
+		opts.VMCtx.Logger.V(4).Info("Skipping PVC disk data backup as unchanged")
 	} else {
 		ecToUpdate = append(ecToUpdate, &types.OptionValue{
-			Key:   vmopv1.VMBackupDiskDataExtraConfigKey,
-			Value: diskDataBackup,
+			Key:   vmopv1.PVCDiskDataExtraConfigKey,
+			Value: pvcDiskData,
 		})
 	}
 
 	if len(ecToUpdate) != 0 {
-		ctx.VMCtx.Logger.Info("Updating VM ExtraConfig with backup data")
-		ctx.VMCtx.Logger.V(4).Info("", "ExtraConfig", ecToUpdate)
-		if _, err := ctx.VcVM.Reconfigure(ctx.VMCtx, types.VirtualMachineConfigSpec{
+		opts.VMCtx.Logger.Info("Updating VM ExtraConfig with latest backup data",
+			"ExtraConfigToUpdate", ecToUpdate)
+		if _, err := opts.VcVM.Reconfigure(opts.VMCtx, types.VirtualMachineConfigSpec{
 			ExtraConfig: ecToUpdate,
 		}); err != nil {
-			ctx.VMCtx.Logger.Error(err, "Failed to update VM ExtraConfig for backup")
+			opts.VMCtx.Logger.Error(err, "failed to update VM ExtraConfig with latest backup data")
 			return err
 		}
 	}
 
 	return nil
 }
 
-func getDesiredVMKubeDataForBackup(
-	vm *vmopv1.VirtualMachine,
+// getDesiredResourceYAMLForBackup returns the encoded and gzipped YAML of the
+// given resources, or an empty string if the data is unchanged in ExtraConfig.
+func getDesiredResourceYAMLForBackup(
+	resources []client.Object,
+	ecKey string,
 	ecMap map[string]string) (string, error) {
-	// If the ExtraConfig already contains the latest VM spec, determined by
-	// 'metadata.generation', return an empty string to skip the backup.
-	if ecKubeData, ok := ecMap[vmopv1.VMBackupKubeDataExtraConfigKey]; ok {
-		vmFromBackup, err := constructVMObj(ecKubeData)
-		if err != nil {
-			return "", err
-		}
-		if vmFromBackup.ObjectMeta.Generation >= vm.ObjectMeta.Generation {
-			return "", nil
+	// Check if the given resources are up-to-date with the latest backup.
+	// This is done by comparing the resource versions of each object UID.
+	var isLatestBackup bool
+	if ecKubeData, ok := ecMap[ecKey]; ok {
+		if resourceToVersion := tryGetResourceVersion(ecKubeData); resourceToVersion != nil {
+			isLatestBackup = true
```

Comment:
nit: Just move this instead of the declaration line 149 so that the logic is clear that by default we assume the backup is latest unless the resource versions do not match.

Reply:
It will not work if the `ecMap` does not contain the backup yet, or if it is somehow corrupted without the resource version info.